lib/sax.js

```diff
@@ -538,86 +538,228 @@ function _copy(source, target) {
+			var comment = getMatch(grammar.Comment);
+			if (comment) {
+				domBuilder.comment(comment, 4, comment.length - 7);
+				return index;
 			} else {
-				//error
-				return -1;
+				return errorHandler.fatalError('Invalid comment starting at position ' + start);
 			}
-		default:
-			if (source.substr(start + 3, 6) == 'CDATA[') {
-				var end = source.indexOf(']]>', start + 9);
+		case '[':
+			// should be CDATA
+			var cdata = getMatch(grammar.CDSect);
+			if (cdata) {
 				domBuilder.startCDATA();
-				domBuilder.characters(source, start + 9, end - start - 9);
+				domBuilder.characters(cdata, 9, cdata.length - 12);
 				domBuilder.endCDATA();
-				return end + 3;
+				return index;
+			} else {
+				return errorHandler.fatalError('Invalid CDATA starting at position ' + start);
 			}
-			//<!DOCTYPE
-			//startDTD(java.lang.String name, java.lang.String publicId, java.lang.String systemId)
-			var matchs = split(source, start);
-			var len = matchs.length;
-			if (len > 1 && /!doctype/i.test(matchs[0][0])) {
-				var name = matchs[1][0];
-				var pubid = false;
-				var sysid = false;
-				if (len > 3) {
-					if (/^public$/i.test(matchs[2][0])) {
-						pubid = matchs[3][0];
-						sysid = len > 4 && matchs[4][0];
-					} else if (/^system$/i.test(matchs[2][0])) {
-						sysid = matchs[3][0];
-					}
+		case 'D': {
+			// should be DOCTYPE
+			var doctype = {
+				name: undefined,
+				publicId: undefined,
+				systemId: undefined,
+				internalSubset: undefined,
+			};
+
+			if (!compare('<!DOCTYPE')) {
+				return errorHandler.fatalError('Expected <!DOCTYPE');
+			}
+			skip(9);
+			if (skipBlanks() < 1) {
+				return errorHandler.fatalError('No blanks between DOCTYPE and document name');
+			}
+
+			// Parse the DOCTYPE name
+			doctype.name = getMatch(grammar.Name);
+			if (!doctype.name) return errorHandler.fatalError('DOCTYPE without name');
+			skipBlanks();
+
+			// Check for External ID and SystemId
+			if (compare('PUBLIC')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between PUBLIC and public identifier');
 				}
-				var lastMatch = matchs[len - 1];
-				domBuilder.startDTD(name, pubid, sysid);
-				domBuilder.endDTD();
+				doctype.publicId = getMatch(grammar.PubidLiteral);
+				if (!doctype.publicId) return errorHandler.fatalError('PUBLIC without public identifier');
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between public identifier and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('PUBLIC without SystemId');
+			} else if (compare('SYSTEM')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between SYSTEM and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('SYSTEM without SystemId');
+			}
+
+			// Parse internal subset
```

Comment:
The only drawback i see with merging the parsing of the whole doctype into the single function is the exploding complexity of the method, which makes it harder/more cumbersome to test.

Reply:
Do you think it would be more readable to make a `DCCParser` class, and move `switch (char(2)) { ... }` into a `parse()` function?

IMO ideally the entire sax parser should be a class, so it can keep track of source, string etc, without them having to be passed around.

Reply:
the way classes are created in the code base in the ES5 way were there is no class keyword, I would prefer to switch to classes and especially not in this PR.

But we could put this whole "index" and those helper functions into a function that returns an object containing all the functions and a `getIndex`, so it can be used in mutliple places.
It's very close to a class, but I don't see the benefit of the whole `prototype` and `new` stuff.

Alternatively there is the functional approach: Every of the functions that currently modify the index instead receive it as an argument and return the new index. (match could no longer contain skip since it needs to return the match).
I think that would be my preferred solution.

Reply:
To be honest I have intentionally refactored this into this pseudo-class style, so it does not require passing the xml and position into every function.
I would very much prefer to leave it like this.
In the event of migrating to ES6 or TS, this code will be easier to port IMO.